src/Toml/Parser/Value.hs

```diff
@@ -41,15 +45,33 @@ decimalP = zero <|> more
     check :: Maybe Integer -> Parser Integer
     check = maybe (fail "Not an integer") pure
 
+-- | Parser for hexadecimal, octal and binary numbers : included parsing
+numberP :: Parser Integer -> Parser Char -> Parser Integer
+numberP parser parseDigit = more
+  where
+    more :: Parser Integer
+    more = check =<< intValueMaybe . concat <$> sepBy1 (some parseDigit) (char '_')
+
+    intValueMaybe :: String -> Maybe Integer
+    intValueMaybe = parseMaybe (parser :: Parser Integer) . fromString
+
+    check :: Maybe Integer -> Parser Integer
+    check = maybe (fail "Not valid number") pure
```

Comment:
You can pass an additional `String` to `numberP` parse to improve this error message, so it will become `"Invalid hexadecimal number"` or `"Invalid binary number"`